GVRf/Framework/framework/src/main/jni/objects/lightlist.cpp

```diff
@@ -0,0 +1,372 @@
+{
+    std::lock_guard < std::recursive_mutex > lock(mLock);
+    bool dirty = mDirty != 0;
+    ShadowMap* shadowMap = NULL;
+
+    if (mDirty & REBUILD_SHADERS)
+    {
+        return scanLights();
+    }
+    if (mDirty & LIGHT_ADDED)
+    {
+        createLightBlock(renderer);
+    }
+    mTotalUniforms = 0;
+    if (mClassMap.size() == 0)
+    {
+        return NULL;
+    }
+    for (auto it1 = mClassMap.begin();
+         it1 != mClassMap.end();
+         ++it1)
+    {
+        const std::vector<Light*>& lights = it1->second;
+        for (auto it2 = lights.begin(); it2 != lights.end(); ++it2)
+        {
+            Light* light = *it2;
+            ShadowMap* sm = light->getShadowMap();
+
+            mTotalUniforms += light->uniforms().uniforms().getNumEntries();
+            if (sm && sm->enabled())
+            {
+                shadowMap = sm;
+            }
+            if (dirty || light->uniforms().isDirty(ShaderData::MAT_DATA))
+            {
+                int offset = light->getBlockOffset();
+                const UniformBlock& uniforms = light->uniforms().uniforms();
+                mLightBlock->updateGPU(renderer, offset, uniforms);
+                light->uniforms().clearDirty();
+            }
+        }
+    }
+    mDirty = 0;
+    return shadowMap;
+}
+
+void LightList::useLights(Renderer* renderer, Shader* shader)
+{
+    if (mUseUniformBlock)
+    {
+        mLightBlock->bindBuffer(shader, renderer);
+    }
+    else
+    {
+        shader->bindLights(*this, renderer);
+    }
+}
+
+void LightList::makeShadowMaps(Scene* scene, jobject jscene, ShaderManager* shaderManager)
+{
+    std::lock_guard < std::recursive_mutex > lock(mLock);
+    int layerIndex = 0;
+    int numShadowMaps = 0;
+
+    for (auto it2 = mClassMap.begin(); it2 != mClassMap.end(); ++it2)
+    {
+        const std::vector<Light*>& lights = it2->second;
+        for (auto it = lights.begin(); it != lights.end(); ++it)
+        {
+            Light *l = (*it);
+            if (l->enabled())
+            {
+                if (l->makeShadowMap(scene, jscene, shaderManager, layerIndex))
+                {
+                    ++numShadowMaps;
+                    ++layerIndex;
+                }
+            }
+        }
+    }
+    if (mNumShadowMaps != numShadowMaps)
+    {
+        mNumShadowMaps = numShadowMaps;
+#ifdef DEBUG_LIGHT
+        LOGD("LIGHT: %d shadow maps", mNumShadowMaps);
+#endif
+    }
+}
+
+bool LightList::createLightBlock(Renderer* renderer)
+{
+    int numFloats = 0;
+
+    for (auto it = mClassMap.begin(); it != mClassMap.end(); ++it)
+    {
+        const std::vector<Light*>& lights = it->second;
+        for (auto it2 = lights.begin(); it2 != lights.end(); ++it2)
+        {
+            Light *light = (*it2);
+            light->setBlockOffset(numFloats);
+            numFloats += light->getTotalSize() / sizeof(float);
+        }
+    }
+    if ((mLightBlock == NULL) ||
+        (numFloats > mLightBlock->getTotalSize()))
```

Comment:
mLightBlock->getTotalSize() should be divided by sizeof(float)

Reply:
@NolaDonato ?

Reply:
Yes you are correct, total size is the number of bytes